pairing_bot.go

```diff
@@ -25,86 +28,84 @@ type PairingLogic struct {
 	adb   APIAuthDB
 	pdb   PairingsDB
 	revdb ReviewDB
-	ur    userRequest
-	un    userNotification
-	sm    streamMessage
 	rcapi RecurseAPI
+
+	zulip *zulip.Client
 }
 
 var randSrc = rand.New(rand.NewSource(time.Now().UnixNano()))
 
 func (pl *PairingLogic) handle(w http.ResponseWriter, r *http.Request) {
 	var err error
 
-	responder := json.NewEncoder(w)
-
 	// check and authorize the incoming request
 	// observation: we only validate requests for /webhooks, i.e. user input through zulip
 
 	ctx := r.Context()
 
 	log.Println("Handling a new Zulip request")
 
-	if err = pl.ur.validateJSON(r); err != nil {
-		log.Println(err)
-		http.NotFound(w, r)
-	}
-
 	botAuth, err := pl.adb.GetKey(ctx, "botauth", "token")
 	if err != nil {
 		log.Println("Something weird happened trying to read the auth token from the database")
 	}
 
-	if !pl.ur.validateAuthCreds(botAuth) {
-		http.NotFound(w, r)
+	hook, err := zulip.ParseWebhook(r.Body, botAuth)
+	if err != nil {
+		log.Println(err)
+		http.NotFound(w, r) // TODO(@jdkaplan): 401 Unauthorized if token mismatch?
+		return
 	}
 
-	intro := pl.ur.validateInteractionType()
-	if intro != nil {
-		if err = responder.Encode(intro); err != nil {
+	// Respond to all public messages with an introduction. Don't process any
+	// commands in open streams/channels.
+	if hook.Trigger != "direct_message" {
+		if err := json.NewEncoder(w).Encode(introMessage); err != nil {
 			log.Println(err)
 		}
 		return
 	}
 
-	ignore := pl.ur.ignoreInteractionType()
-	if ignore != nil {
-		if err = responder.Encode(ignore); err != nil {
+	// Don't respond to commands sent in pair-making group DMs. If there aren't
+	// exactly two participants (Pairing Bot + 1), keep quiet!
+	//
+	// TODO(@jdkaplan): Maybe ACK the message and point the sender elsewhere?
```

Comment:
On second thought, this would mean an extra notification to all members of the group. Leave it as-is!
```suggestion
```